crates/matrix-sdk-ui/src/timeline/event_item/mod.rs

```diff
@@ -693,3 +693,75 @@ impl ReactionsByKeyBySender {
         None
     }
 }
+
+/// Extends [`ShieldState`] to allow for a `SentInClear` code.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum TimelineEventShieldState {
+    /// A red shield with a tooltip containing the associated message should be
+    /// presented.
+    Red {
+        /// A machine-readable representation.
+        code: TimelineEventShieldStateCode,
+        /// A human readable description.
+        message: &'static str,
+    },
+    /// A grey shield with a tooltip containing the associated message should be
+    /// presented.
+    Grey {
+        /// A machine-readable representation.
+        code: TimelineEventShieldStateCode,
+        /// A human readable description.
+        message: &'static str,
+    },
+    /// No shield should be presented.
+    None,
+}
+
+impl From<ShieldState> for TimelineEventShieldState {
+    fn from(value: ShieldState) -> Self {
+        match value {
+            ShieldState::Red { code, message } => {
+                TimelineEventShieldState::Red { code: code.into(), message }
+            }
+            ShieldState::Grey { code, message } => {
+                TimelineEventShieldState::Grey { code: code.into(), message }
+            }
+            ShieldState::None => TimelineEventShieldState::None,
+        }
+    }
+}
+
+/// Extends [`ShieldStateCode`] to allow for a `SentInClear` code.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
+pub enum TimelineEventShieldStateCode {
```

Comment:
I quite like how [errors derive their display implementation](https://github.com/dtolnay/thiserror), I wonder if there's anything like that for normal enums. Then again, any reason why `TimelineEventShieldStateCode` can't be an actual `thiserror::Error`?